### PR TITLE
fix: Optimizes GetString to eliminate allocations

### PR DESCRIPTION
**Pull request by kamronbatman**
### Summary

* Optimized GetString by eliminating the intermediate string allocation.

** Note **: Encoding.GetChars() is still really inefficient, especiall when strings are not aligned or have regular ascii/unicode characters. Thankfully we generally don't have to worry about these odd edge cases, but if they happen then GetChars can allocate hundreds of bytes.

This is a benchmark for just the related changes. NonSpecial are just ascii characters, while the other tests include control codes.
```cs
| Method                         | Mean     | Error   | StdDev  | Gen0   | Allocated |
|------------------------------- |---------:|--------:|--------:|-------:|----------:|
| GetString                      | 306.7 ns | 5.96 ns | 6.86 ns | 0.0124 |     200 B |
| GetStringNotSpecial            | 257.0 ns | 4.83 ns | 4.52 ns | 0.0114 |     184 B |
| GetStringSpanHelpers           | 166.4 ns | 3.26 ns | 3.48 ns |      - |         - |
| GetStringSpanHelpersNotSpecial | 137.3 ns | 1.57 ns | 1.22 ns |      - |         - |
```

### CODE REVIEW

File: Projects/Server/Text/TextEncoding.cs [hunk 1]
```diff
@@ -114,33 +115,55 @@ public static int GetByteLengthForEncoding(this Encoding encoding) =>
             _          => 1
         };
 
-    [MethodImpl(MethodImplOptions.AggressiveInlining)]
-    private static bool IsSafeChar(ushort c) => c is >= 0x20 and < 0xFFFE;
-
     public static string GetString(ReadOnlySpan<byte> span, Encoding encoding, bool safeString = false)
     {
-        string s = encoding.GetString(span);
```
Comment (kamronbatman) [inline Projects/Server/Text/TextEncoding.cs:122]: Honestly this is the crux of the change.

File: Projects/Server/Text/TextEncoding.cs [hunk 2]
```diff
@@ -114,33 +115,55 @@ public static int GetByteLengthForEncoding(this Encoding encoding) =>
...
-
         if (!safeString)
         {
-            return s;
+            return encoding.GetString(span);
         }
 
-        ReadOnlySpan<char> chars = s.AsSpan();
+        var charCount = encoding.GetMaxCharCount(span.Length);
```
Comment (kamronbatman) [inline Projects/Server/Text/TextEncoding.cs:125]: We use GetMaxCharCount instead of GetCharCount since the latter is considerably slower and sometimes allocates.

File: Projects/Server/Text/TextEncoding.cs [hunk 3]
```diff
@@ -114,33 +115,55 @@ public static int GetByteLengthForEncoding(this Encoding encoding) =>
...
+        Span<char> chars = charCount <= 256
+            ? stackalloc char[charCount]
+            : rentedChars = STArrayPool<char>.Shared.Rent(charCount);
 
-        for (int i = 0, last = 0; i < chars.Length; i++)
+        try
         {
-            if (!IsSafeChar(chars[i]) && i == chars.Length - 1)
+            var length = encoding.GetChars(span, chars);
+            chars = chars[..length];
```
Comment (kamronbatman) [inline Projects/Server/Text/TextEncoding.cs:135]: This still allocates with weird strings. I am not going to spend the time to guard against malicious input that could cause 2-5x perf hit or the 300+ byte allocation. It's unfortunate, but maybe someone knows of a way to check/sanitize a buffer for real characters and effectively bail out early if the string is corrupt/nonsense.
